packages/ui/src/views/repo/pull-request/details/components/conversation/regular-and-code-comment.tsx

```diff
@@ -0,0 +1,340 @@
+import { FC, useCallback, useMemo, useState } from 'react'
+
+import { Avatar, Icon, Layout } from '@/components'
+import {
+  CommentItem,
+  isCodeComment,
+  PullRequestCommentBox,
+  PullRequestOverviewProps,
+  removeLastPlus,
+  TranslationStore,
+  TypesPullReqActivity
+} from '@/views'
+import { DiffModeEnum } from '@git-diff-view/react'
+import { getInitials } from '@utils/stringUtils'
+import { timeAgo } from '@utils/utils'
+import PullRequestDiffViewer from '@views/repo/pull-request/components/pull-request-diff-viewer'
+import { PRCommentViewProps } from '@views/repo/pull-request/details/components/common/pull-request-comment-view'
+import PullRequestTimelineItem, {
+  TimelineItemProps
+} from '@views/repo/pull-request/details/components/conversation/pull-request-timeline-item'
+import { useDiffConfig } from '@views/repo/pull-request/hooks/useDiffConfig'
+import { parseStartingLineIfOne, quoteTransform } from '@views/repo/pull-request/utils'
+import { get } from 'lodash-es'
+
+const getAvatar = (name?: string) => {
+  return (
+    <Avatar.Root>
+      <Avatar.Fallback>{getInitials(name || '')}</Avatar.Fallback>
+    </Avatar.Root>
+  )
+}
+
+export interface PullRequestRegularAndCodeCommentProps
+  extends Pick<
+    PullRequestOverviewProps,
+    | 'handleUpload'
+    | 'currentUser'
+    | 'toggleConversationStatus'
+    | 'handleSaveComment'
+    | 'onCopyClick'
+    | 'handleDeleteComment'
+    | 'handleUpdateComment'
+  > {
+  commentItems: CommentItem<TypesPullReqActivity>[]
+  isLast: boolean
+  useTranslationStore: () => TranslationStore
+  componentViewBase: FC<{ commentItem: PRCommentViewProps['commentItem'] }>
+}
+
+export const PullRequestRegularAndCodeComment: FC<PullRequestRegularAndCodeCommentProps> = ({
+  commentItems,
+  handleUpload,
+  currentUser,
+  toggleConversationStatus,
+  isLast,
+  handleSaveComment,
+  onCopyClick,
+  handleDeleteComment,
+  useTranslationStore,
+  handleUpdateComment,
+  componentViewBase: ComponentViewBase
+}) => {
+  const { t } = useTranslationStore()
+  const { highlight, wrap, fontsize } = useDiffConfig()
+
+  const [hideReplyHeres, setHideReplyHeres] = useState<{ [key: string]: boolean }>({})
+  const [quoteReplies, setQuoteReplies] = useState<Record<number, { text: string }>>({})
+  const [editModes, setEditModes] = useState<{ [key: string]: boolean }>({})
+  const [editComments, setEditComments] = useState<{ [key: string]: string }>({})
+
+  const payload = useMemo(() => {
+    return commentItems[0]?.payload
+  }, [commentItems])
```

Comment:
Probably no need for a `useMemo` here. It won't stabilise the reference and the overhead will slow things down a wee bit.
```suggestion
  const payload = commentItems[0]?.payload
```

Reply:
In that case, I wrapped the entire component in `memo` since payload is currently used as a dependency in multiple hooks. This way, I can remove all individually memoized data, and the component itself will only re-render upon receiving new props.